frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx

```diff
@@ -263,50 +241,150 @@ function DataFrame({
+    isColumnSelected,
+    isCellSelected,
+    clearSelection,
+    clearCellSelection,
+    processSelectionChange,
+  } = useSelectionHandler(element, isEmptyTable, disabled, syncSelectionState)
+
+  // This callback is used to refresh the rendering of specified cells
+  const refreshCells = React.useCallback(
+    (
+      cells: {
+        cell: [number, number]
+      }[]
+    ) => {
+      dataEditorRef.current?.updateCells(cells)
+    },
+    []
+  )
 
-        if (currentWidgetState === undefined) {
-          // Create an empty widget state
-          currentWidgetState = new EditingState(0).toJson([])
-        }
+  /**
+   * This callback is used to update the number of rows based
+   * on the latest editing state. This is required to keep the
+   * component state in sync with the editing state.
+   */
+  const updateNumRows = React.useCallback(() => {
+    if (numRows !== editingState.current.getNumRows()) {
+      // Reset the number of rows if it has been changed in the editing state
+      setNumRows(editingState.current.getNumRows())
+    }
+  }, [numRows])
 
-        // Only update if there is actually a difference between editing and widget state
-        if (currentEditingState !== currentWidgetState) {
-          widgetMgr.setStringValue(
-            element as WidgetInfo,
-            currentEditingState,
-            {
-              fromUi: triggerRerun,
-            },
-            fragmentId
-          )
-        }
-      })()
-    },
-    [widgetMgr, element, numRows, clearSelection, columns]
+  /**
+   * This callback is used to synchronize the editing state with
+   * the widget state of the component. This might also send a rerun message
+   * to the backend if the editing state has changed.
+   */
+  // The debounce method doesn't allow dependency inspection. Therefore, we
+  // need to disable the eslint rule for exhaustive-deps.
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const syncEditState = React.useCallback(
+    // Use debounce to prevent rapid updates to the widget state.
+    debounce(DEBOUNCE_TIME_MS, () => {
+      const currentEditingState = editingState.current.toJson(columns)
+      let currentWidgetState = widgetMgr.getStringValue({
+        id: element.id,
+        formId: element.formId,
+      } as WidgetInfo)
+
+      if (currentWidgetState === undefined) {
+        // Create an empty widget state
+        currentWidgetState = new EditingState(0).toJson([])
+      }
+
+      // Only update if there is actually a difference between editing and widget state
+      if (currentEditingState !== currentWidgetState) {
+        widgetMgr.setStringValue(
+          {
+            id: element.id,
+            formId: element.formId,
+          } as WidgetInfo,
+          currentEditingState,
+          {
+            fromUi: true,
+          },
+          fragmentId
+        )
+      }
+    }),
+    [
+      element.id,
+      element.formId,
+      widgetMgr,
+      element,
```

Comment:
good point, yep `element` needs to be removed. `element` is anyways not working correctly as dependency since its an object that might change its identity.